src/mlpack/core/dists/diag_cov_gaussian_distribution.hpp

```diff
@@ -0,0 +1,210 @@
+
+  /**
+   * Calculates the inverse of the given diagonal covariance.
+   */
+  void InverseCovariance();
+
+  /**
+   * Return a randomly generated observation according to the probability
+   * distribution defined by this object.
+   * 
+   * @ return Random observation from this Diagonal Gaussian distribution.
+   */
+  arma::vec Random() const;
+
+  /**
+   * Estimate the Gaussian distribution directly from the given observations.
+   * 
+   * @param observations List of observations.
+   */
+  void Train(const arma::mat& observations);
+
+  /**
+   * Estimate the Gaussian distribution from the given observations,
+   * taking into account the probability of each observation actually being
+   * from this distribution.
+   * 
+   * @param observations Matrix of observations.
+   * @param probabilities List of probability of the each observation being
+   * from this distribution.
+   */
+  void Train(const arma::mat& observations,
+             const arma::vec& probabilities);
+
+  /**
+   * Return the mean.
+   */
+  const arma::vec& Mean() const { return mean; }
+
+  /**
+   * Return a modifiable copy of the mean.
+   */
+  arma::vec& Mean() { return mean; }
+
+  /**
+   * Return the covariance matrix.
+   */
+  const arma::mat& Covariance() const { return covariance; }
+
+  /**
+   * Set the covariance.
+   */
+  void Covariance(const arma::mat& covariance);
+
+  void Covariance(arma::mat&& covariance);
+
+  /**
+   * Serialize the distribution.
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */)
+  {
+    // We just need to serialize each of the members.
+    ar & BOOST_SERIALIZATION_NVP(mean);
+    ar & BOOST_SERIALIZATION_NVP(covariance);
+    ar & BOOST_SERIALIZATION_NVP(invCov);
+    ar & BOOST_SERIALIZATION_NVP(logDetCov);
+  }
+};
+
+/**
+* Calculates the multivariate Gaussian log probability density function for
+* each data point (column) in the given matrix.
+*
+* @param observations Matrix of observations.
+* @param probabilities Output log probabilities for each input observation.
+*/
+inline void DiagCovGaussianDistribution::LogProbability(
+    const arma::mat& observations,
+    arma::vec& logProbabilities) const
+{
+  // Column i of 'diffs' is the difference between observations.col(i) and
+  // the mean.
+  arma::mat diffs = observations -
+      (mean * arma::ones<arma::rowvec>(observations.n_cols));
+
+  // Now, we only want to calculate the diagonal elements of (diffs' * cov^-1 *
+  // diffs).  We just don't need any of the other elements.  We can calculate
+  // the right hand part of the equation (instead of the left side) so that
+  // later we are referencing columns, not rows -- that is faster.
+  const size_t k = observations.n_rows;
+  const arma::mat rhs = -0.5 * arma::diagmat(invCov) * diffs;
+  arma::vec logExponents(diffs.n_cols);
+
+  for (size_t i = 0; i < diffs.n_cols; i++)
+    logExponents(i) = accu(diffs.unsafe_col(i) % rhs.unsafe_col(i));
```

Comment:
I think that can work, but I'd suggest holding `invCov` as an `arma::vec`, then doing

```
arma::vec logExponents = -0.5 * arma::pow(diffs, 2).t() * arma::diagmat(invCov);
```

Reply:
Sorry for confusing you. The `Equation` is a example about a single observation like `arma::vec`, but the `Code` is example about multiple observations like `arma::mat`. So, in case of multiple observation(column-wise), Isn't the code correct??

Reply:
Oh, right, sorry!  If you're looking to do elementwise multiplication, shouldn't we use the `%` operator instead of `*`?  i.e. `arma::pow(diffs, 2).t() % invCov`.